### PR TITLE
Starting view

### PR DESCRIPTION
**Pull request by AniiMancuso**
## Summary ##

**Trello card**: https://trello.com/c/zsHE9jW8

The Login and part of the library structure have been created

### iPhone SE

Captura de Pantalla 2022-09-21 a la(s) 14 32 43

### iPhone 14 Pro Max

Captura de Pantalla 2022-09-21 a la(s) 14 31 42

## Known issues ##

I have had some difficulties in understanding the errors that the program presented and in my spare time I had progressed far beyond the login, that is why in this pull request you can see how in addition to designing the login card, I have started with the cards of the design of the library and that of the navigation bar.

### CODE REVIEW

File: WBooks/Library.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        guard let name = aDecoder.decodeObject(forKey: PropertyKey.name) as? String else {
+            os_log("Unable to decode the name for a Library object.", log: OSLog.default, type: .debug)
+            return nil
+        }
+        
+        // Because photo is an optional property of Meal, just use conditional cast.
+        let photo = aDecoder.decodeObject(forKey: PropertyKey.photo) as? UIImage
+        
+        guard let name2 = aDecoder.decodeObject(forKey: PropertyKey.name2) as? String else {
+            os_log("Unable to decode the name2 for a Library object.", log: OSLog.default, type: .debug)
```
Comment (dfmm472) [inline WBooks/Library.swift:46]: change for debugPrint

File: WBooks/Library.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
+//
+//  Library.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 20/09/2022.
+//
+
+import UIKit
+import os.log
```
Comment (dfmm472) [inline WBooks/Library.swift:9]: not needed if debugPrint is used

File: WBooks/Library.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        static let name2 = "name2"
+    }
+    
+    //MARK: Initialization
+     
+    required convenience init?(coder aDecoder: NSCoder) {
+        
+        // The name is required. If we cannot decode a name string, the initializer should fail.
+        guard let name = aDecoder.decodeObject(forKey: PropertyKey.name) as? String else {
+            os_log("Unable to decode the name for a Library object.", log: OSLog.default, type: .debug)
```
Comment (dfmm472) [inline WBooks/Library.swift:38]: change for debugPrint

File: WBooks/Library.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        
+        // Must call designated initializer.
+        self.init(name: name, photo: photo, name2: name2)
+        
+    }
+    init?(name: String, photo: UIImage?, name2: String) {
+        // The name must not be empty
+        guard !name.isEmpty else {
+            return nil
```
Comment (dfmm472) [inline WBooks/Library.swift:57]: I think the return nil can be left just return

File: WBooks/Library.swift [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    //MARK: Properties
+    
+    var name: String
+    var photo: UIImage?
+    var name2: String
+    
+    //MARK: Archiving Paths
+     
+    static let DocumentsDirectory = FileManager().urls(for: .documentDirectory, in: .userDomainMask).first!
+    static let ArchiveURL = DocumentsDirectory.appendingPathComponent("library")
```
Comment (dfmm472) [inline WBooks/Library.swift:22]: are they used somewhere?

File: WBooks/Library.swift [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+        
+    //MARK: NSCoding
+    func encode(with aCoder: NSCoder) {
+        aCoder.encode(name, forKey: PropertyKey.name)
+        aCoder.encode(photo, forKey: PropertyKey.photo)
+        aCoder.encode(name2, forKey: PropertyKey.name2)
+        
+    }
+}
```
Comment (dfmm472) [inline WBooks/Library.swift:74]: add line break, since the class started with line break

File: WBooks/LibraryTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    @IBOutlet weak var nameLabel2: UILabel!
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+
```
Comment (dfmm472) [inline WBooks/LibraryTableViewCell.swift:25]: delete extra line break

File: WBooks/Library.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        
+        guard let name2 = aDecoder.decodeObject(forKey: PropertyKey.name2) as? String else {
+            os_log("Unable to decode the name2 for a Library object.", log: OSLog.default, type: .debug)
+            return nil
+        }
+        
+        // Must call designated initializer.
+        self.init(name: name, photo: photo, name2: name2)
+        
+    }
```
Comment (dfmm472) [inline WBooks/Library.swift:53]: add line break

File: WBooks/LibraryTableViewController.swift [hunk 9]
```diff
@@ -0,0 +1,116 @@
+//
+//  LibraryTableViewController.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 20/09/2022.
+//
+
+import UIKit
+import os.log
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:9]: same

File: WBooks/LibraryTableViewController.swift [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    //MARK: Properties
+    var library = [Library]()
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        navigationItem.leftBarButtonItem = editButtonItem
+        if let savedLibrary = loadLibrary() {
+            library += savedLibrary
+        }
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:21]: delete line break

File: WBooks/LibraryTableViewController.swift [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        let librarys = library[indexPath.row]
+        
+        cell.nameLabel.text = librarys.name
+        cell.photoImageView.image = librarys.photo
+        cell.nameLabel2.text = librarys.name2
+
+        return cell
+    }
+    
+
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:55]: delete line break

File: WBooks/LibraryTableViewController.swift [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    // Override to support conditional editing of the table view.
+    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+        // Return false if you do not want the specified item to be editable.
+        return true
+    }
+
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:61]: delete line break

File: WBooks/LibraryTableViewController.swift [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+             fatalError("Unable to instantiate cover5")
+         }
+         
+         library += [cover1, cover2, cover3, cover4, cover5]
+         
+     }
+     private func saveLibrary() {
+         let isSuccessfulSave = NSKeyedArchiver.archiveRootObject(library, toFile: Library.ArchiveURL.path)
+         if isSuccessfulSave {
+             os_log("Library successfully saved.", log: OSLog.default, type: .debug)
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:108]: change for debugPrint

File: WBooks/LibraryTableViewController.swift [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+         } else {
+             os_log("Failed to save Library...", log: OSLog.default, type: .error)
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:110]: same

File: WBooks/LibraryTableViewController.swift [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+         }
+     }
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:112]: add break line

File: WBooks/ViewController.swift [hunk 16]
```diff
@@ -9,11 +9,19 @@ import UIKit
...
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
     }
-
-
+    
+    //MARK: Action
+    @IBAction func singinButton(_ sender: UIButton) {
+        print("Boton ingreso pulsado")
+    }
```
Comment (dfmm472) [inline WBooks/ViewController.swift:25]: add break line

File: WBooks/LibraryTableViewController.swift [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+     private func loadSampleLibrarys() {
+         
+         let photo1 = UIImage(named: "cover1")
+         let photo2 = UIImage(named: "cover2")
+         let photo3 = UIImage(named: "cover3")
+         let photo4 = UIImage(named: "cover4")
+         let photo5 = UIImage(named: "cover5")
+         
+         guard let cover1 = Library(name: "A Little Bird Told Me", photo: photo1, name2: "Timothy Cross") else {
+             fatalError("Unable to instantiate cover1")
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:86]: we can change that fatal error by a logic in which if the image is not there we do not show it or we put one by default

File: WBooks/LibraryTableViewController.swift [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+             fatalError("Unable to instantiate cover2")
+         }
+         guard let cover3 = Library(name: "The Best Book in the World", photo: photo3, name2: "Peter Sjernstrom") else {
+             fatalError("Unable to instantiate cover3")
+         }
+         guard let cover4 = Library(name: "Be Creative", photo: photo4, name2: "Tony Alcazar") else {
+             fatalError("Unable to instantiate cover4")
+         }
+         guard let cover5 = Library(name: "Redesign The Web", photo: photo5, name2: "Liliana Castilla") else {
+             fatalError("Unable to instantiate cover5")
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:99]: same as comment above

File: WBooks/LibraryTableViewController.swift [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+         }
+         
+         library += [cover1, cover2, cover3, cover4, cover5]
+         
+     }
```
Comment (dfmm472) [inline WBooks/LibraryTableViewController.swift:104]: add break line

File: WBooks/Library.swift (hunk 4 above)
Comment (AniiMancuso) [reply WBooks/Library.swift:57]: I tried if it could be done with the return alone but it didn't let me, it gave me an error in the self.name

**Comment by dfmm472** [comment]
The adjustments are made in the Main Menu branch, since it was not clear what they should be done in this same branch, and they were started in the other

**Review by dfmm472 — APPROVED** [review]

**Pull request merged**